aten/src/ATen/native/Copy.cpp

```diff
@@ -69,6 +72,14 @@ void copy_same_type_transpose_(Tensor& self, const Tensor& src) {
       }
     }
   });
+#ifdef BUILD_NAMEDTENSOR
+  auto outnames = unify_from_right(self.names(), src.names());
+  if (outnames.has_value()) {
+    at::internal_set_names_inplace(self, *outnames);
+  } else {
+    at::internal_set_names_inplace(self, nullopt);
```

Comment:
if `!outnames.has_value()`, isn't it already a nullopt?

Reply:
as in, it seems like this could just be `at::internal_set_names_inplace(self, unify_from_right(self.names(), src.names());` ?

Reply:
No, it can't be. The problem is that `outnames` is an `optional<vector<Dimname>>`, while `internal_set_names_inplace` takes `optional<ArrayRef<Dimname>>` and one can't pass a `optional<vector<Dimname>>` for a `optional<ArrayRef<Dimname>>`.

I can add an extra overload for `at::internal_set_names_inplace` to make it nicer.

Reply:
just to make sure I understand that correctly:

You can pass a `vector<Dimname>` to `ArrayRef<Dimname>`, but not `optional<vector<Dimname>>` to `optional<ArrayRef<Dimname>>`?

Is it the array ref constructor (`ArrayRef(const std::vector<T, A>& Vec)`) that makes the former possible? 

For it to work with optional, there'd need to be some kind of special optional constructor that took one type and constructed an optional of a different type using that constructor?

Reply:
> You can pass a vector<Dimname> to ArrayRef<Dimname>, but not optional<vector<Dimname>> to optional<ArrayRef<Dimname>>?

Yes

> Is it the array ref constructor (ArrayRef(const std::vector<T, A>& Vec)) that makes the former possible?

Yes. That's because it is an implicit constructor.

> For it to work with optional, there'd need to be some kind of special optional constructor that took one type and constructed an optional of a different type using that constructor?

I don't know if that's possible in C++, but yes there would need to be something like that. Alternative, we'd need something like a `bind` operator that takes a `optional<T>` and a `f: T -> optional<U>` and returns `optional<U>`, but that's not in the c++ optional spec.

Reply:
Oh, one thing that's interesting to note is that one can also pass a `vector` to an `optional<ArrayRef>`.  I believe what lets this happen is also the implicit constructor `ArrayRef(const std::vector<T, A>& Vec))` (we implicitly convert vector -> ArrayRef and then there is an optional<T> constructor that takes a T.

Reply:
interesting, thanks